File: c++ version/101-200/113-path-sum-II.cpp
class Solution {
    vector<vector<int>> ans;
    vector<int> path;
public:
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        if (root == nullptr) return ans;
        backtracking(root, targetSum);
        return ans;
    }

    void backtracking(TreeNode* root, int target) {
        if (root == nullptr) return;
        path.push_back(root->val);
        if (!root->left && !root->right) {
            if (target == root->val) {
                ans.push_back(path);
            }
        }
        backtracking(root->left, target-root->val);
        backtracking(root->right, target-root->val);
        path.pop_back();
    }
};

File: python version/113-path-sum-II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        ans = []
        curr = []

        def dfs_helper(root, target):
            if not root:
                return
            curr.append(root.val)
            target -= root.val
            if not root.left and not root.right:
                if target == 0:
                    ans.append(curr[:])
            dfs_helper(root.left, target)
            dfs_helper(root.right, target)
            curr.pop()
        
        dfs_helper(root, targetSum)
        return ans
